### PR TITLE
Fix issue in optional field access 

### LINKED ISSUE #25673: Chain optional field access fails with a NPE
**Description:**
Chain optional field access fails for following code-snippet.

```ballerina
public function main() {
    Person person = new Person();
    string? c = person.getDetails()?.addr?.street;
}

public type Person object {
    public function getDetails() returns Details? {
        return {};
    }
};

public type Details record {
    Address addr?;
};

public type Address record {
    string street;
};
```

**Error Thrown**
```
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.desugar.Desugar.handleSafeNavigation(Desugar.java:6388)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.handleSafeNavigation(Desugar.java:6383)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.handleSafeNavigation(Desugar.java:6383)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewriteSafeNavigationExpr(Desugar.java:6356)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:3391)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangFieldBasedAccess.accept(BLangFieldBasedAccess.java:65)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewriteExpr(Desugar.java:5298)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1041)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5273)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1162)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5273)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5314)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewriteStmt(Desugar.java:5324)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:964)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5273)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1002)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5273)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5331)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:694)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:153)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:5273)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.perform(Desugar.java:381)
	at org.wso2.ballerinalang.compiler.CompilerDriver.desugar(CompilerDriver.java:350)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:296)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:248)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:140)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:166)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:99)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:57)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:299)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)
```

**Affected Versions:**
2.0.0-P4-SNAPSHOT

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Fix an issue occurred in optional field access for method invocation
$subject

Fixes #25673

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal [hunk 1]
```diff
@@ -1,4 +1,3 @@
-// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (KRVPerera) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal:1]: intended ?

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal:1]: This is a mistake, fixed it

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -5290,7 +5290,7 @@ private BSymbol getLangLibMethod(BLangInvocation iExpr, BType bType) {
     }
 
     private void checkInvocationParamAndReturnType(BLangInvocation iExpr) {
-        BType actualType = checkInvocationParam(iExpr);
+        BType actualType = iExpr.originalType = checkInvocationParam(iExpr);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5293]: While this change fixes the issue, wonder if we can remove the check for `INVOCATION` at https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L7770 altogether. 

IIRC, previously error/nil lifting access was allowed on invocations, but that is not allowed anymore. So we may be able to remove that condition. Shall we check if this is possible?

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal [hunk 3]
```diff
@@ -391,3 +391,40 @@ function testOptionalFieldAccessInUnionType() {
...
+};
+
+public type Address record {
+    string street;
+};
+
+function testOptionalFieldAccessOnClassDef() {
+    Address addr = {
+        street : "Colombo"
+    };
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal]: Fixed.

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal]: Fixed.

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal [hunk 5]
```diff
@@ -391,3 +391,40 @@ function testOptionalFieldAccessInUnionType() {
...
+};
+
+function testOptionalFieldAccessOnClassDef() {
+    Address addr = {
+        street : "Colombo"
+    };
+
+    Details details = {
+        addr : addr
+    };
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal]: Fixed

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal [hunk 6]
```diff
@@ -391,3 +391,40 @@ function testOptionalFieldAccessInUnionType() {
...
+
+    Student person = new Student(details);
+    string? c1 = person.getDetails()?.addr?.street;
+
+    if (c1 != "Colombo") {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal:423]: Added

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 2 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5289]: Changed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 7]
```diff
@@ -7769,8 +7769,7 @@ private void handleSafeNavigation(BLangAccessExpression accessExpr, BType type,
         // If the parent of current expr is the root, terminate
         NodeKind kind = accessExpr.expr.getKind();
         if (kind == NodeKind.FIELD_BASED_ACCESS_EXPR ||
-                kind == NodeKind.INDEX_BASED_ACCESS_EXPR ||
-                kind == NodeKind.INVOCATION) {
+                kind == NodeKind.INDEX_BASED_ACCESS_EXPR) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:7772]: Can merge these 2 lines.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 2 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5289]: We don't need this change anymore, right?

Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Yes, removed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 7 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Done

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal [hunk 8]
```diff
@@ -391,3 +391,43 @@ function testOptionalFieldAccessInUnionType() {
...
+
+public type Details record {
+    Address addr?;
+};
+
+public type Address record {
+    string street;
+};
+
+function testOptionalFieldAccessOnClassDef() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal:415]: Shall we rename this to 
```suggestion
function testOptionalFieldAccessOnMethodCall() {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/access/OptionalFieldAccessTest.java [hunk 9]
```diff
@@ -137,4 +139,9 @@ public void testOptionalFieldAccessOnInvocation(String function) {
     public void testOptionalFieldAccessInUnionType() {
         BRunUtil.invoke(result, "testOptionalFieldAccessInUnionType");
     }
+
+    @Test
+    public void testOptionalFieldAccessOnClassDef() {
+        BRunUtil.invoke(result, "testOptionalFieldAccessOnClassDef");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/access/OptionalFieldAccessTest.java:145]:
```suggestion
    public void testOptionalFieldAccessOnMethodCall() {
        BRunUtil.invoke(result, "testOptionalFieldAccessOnMethodCall");
```

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal (hunk 8 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/optional_field_access.bal]: Fixed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/access/OptionalFieldAccessTest.java (hunk 9 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/access/OptionalFieldAccessTest.java]: Fixed

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**